pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/Node.java

```diff
@@ -224,11 +224,25 @@
      * @param xpathString
      *            the expression to check
      * @return List of all matching nodes. Returns an empty list if none found.
-     * @throws JaxenException
+     * @throws JaxenException if the xpath is incorrect or fails altogether
      */
     List<? extends Node> findChildNodesWithXPath(String xpathString) throws JaxenException;
 
     /**
+     * Returns all the nodes matching the xpath expression and casts the nodes to a specific type.
+     *
+     * @param clazz
+     *            the class of the result component type
+     * @param xpathString
+     *            the expression to check
+     * @return List of all matching nodes. Returns an empty list if none found.
+     *
+     * @throws JaxenException if the xpath is incorrect or fails altogether
+     * @throws ClassCastException if the nodes found with the xpath exception cannot be cast.
+     */
+    <T extends Node> List<T> findChildNodesWithXPathChecked(Class<T> clazz, String xpathString) throws JaxenException;
```

Comment:
I'm unsure, whether we should keep it - since you don't require it anymore.
Having it added into this interface, will make it difficult to change afterwards.
So, I'd vote for removing this method for now.

Note: This method will work for simple XPath expressions, that match a single node type only. But you can easily create expressions, that match multiple types (`ClassOrInterfaceDeclaration | EnumDeclaration`). Ok, for this example, you could use your new `ASTAnyTypeDeclaration.class`.